src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs

```diff
@@ -0,0 +1,431 @@
+            }
+
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataEntry
+            {
+                public fixed byte id[40];
+                public int byteOffset;
+                public int length;
+                public int unused; // Was raw GMT offset; always 0 since tzdata2014f (L).
+            }
+
+            private string[]? _ids;
+            private int[]? _byteOffsets;
+            private int[]? _lengths;
+            private string _tzFileDir;
+            private string _tzFilePath;
+
+            private static string GetApexTimeDataRoot()
+            {
+                string? ret = Environment.GetEnvironmentVariable("ANDROID_TZDATA_ROOT");
+                if (!string.IsNullOrEmpty(ret))
+                {
+                    return ret;
+                }
+
+                return "/apex/com.android.tzdata";
+            }
+
+            private static string GetApexRuntimeRoot()
+            {
+                string? ret = Environment.GetEnvironmentVariable("ANDROID_RUNTIME_ROOT");
+                if (!string.IsNullOrEmpty(ret))
+                {
+                    return ret;
+                }
+
+                return "/apex/com.android.runtime";
+            }
+
+            public AndroidTzData()
+            {
+
+                string[] tzFileDirList = new string[] {GetApexTimeDataRoot() + "/etc/tz/", // Android 10+, TimeData module where the updates land
+                                                       GetApexRuntimeRoot() + "/etc/tz/", // Android 10+, Fallback location if the above isn't found or corrupted
+                                                       Environment.GetEnvironmentVariable("ANDROID_DATA") + "/misc/zoneinfo/",
+                                                       Environment.GetEnvironmentVariable("ANDROID_ROOT") + DefaultTimeZoneDirectory};
+                foreach (var tzFileDir in tzFileDirList)
+                {
+                    string tzFilePath = Path.Combine(tzFileDir, TimeZoneFileName);
+                    if (LoadData(tzFilePath))
+                    {
+                        _tzFileDir = tzFileDir;
+                        _tzFilePath = tzFilePath;
+                        return;
+                    }
+                }
+
+                throw new TimeZoneNotFoundException(SR.TimeZoneNotFound_ValidTimeZoneFileMissing);
+            }
+
+            private bool LoadData(string path)
+            {
+                if (!File.Exists(path))
+                {
+                    return false;
+                }
+                try {
+                    using (FileStream fs = File.OpenRead(path))
+                    {
+                        LoadTzFile(fs);
+                    }
+                    return true;
+                }
+                catch (IOException) {}
+
+                return false;
+            }
+
+            private unsafe void LoadTzFile(Stream fs)
```

Comment:
Does this method need to be `unsafe` anymore?

Reply:
No, good catch :)